File: cpp/tictactoe4.cpp
#include <iostream>

const int N = 4;
char field[N][N];

void playerMove()
{
    int x, y;
    do
    {
        std::cout << "Куда будете ходить? ";
        char c;
        std::cin >> c;
        x = c - 'a';
        std::cin >> c;
        y = c - '1';
    }
    while (x < 0 || x >= N || y < 0 || y >= N || field[y][x]);
    field[y][x] = 'X';
}

bool checkLines()
{
    bool found;
    for (int i = 0 ; i < N ; ++i)
    {
        found = field[i][0] != 0;
        for (int j = 1 ; found && j < N ; ++j)
        {
            found = field[i][j] == field[i][0];
        }
        if (found)
        {
            return true;
        }
        found = field[0][i] != 0;
        for (int j = 1 ; found && j < N ; ++j)
        {
            found = field[j][i] == field[0][i];
        }
        if (found)
        {
            return true;
        }
    }
    found = field[0][0] != 0;
    for (int j = 1 ; found && j < N ; ++j)
    {
        found = field[j][j] == field[0][0];
    }
    return found;
}

bool canWin(bool player)
{
    if (checkLines())
    {
        return false;
    }
    for (int x = 0 ; x < N ; ++x)
    {
        for (int y = 0 ; y < N ; ++y)
        {
            if (!field[y][x])
            {
                field[y][x] = player ? 'X' : 'O';
                bool w = canWin(!player);
                field[y][x] = 0;
                if (!player && w)
                {
                    return true;
                }
                if (player && !w)
                {
                    return false;
                }
            }
        }
    }
    return player;
}

void computerMove()
{
    for (int x = 0 ; x < N ; ++x)
    {
        for (int y = 0 ; y < N ; ++y)
        {
            if (!field[y][x])
            {
                field[y][x] = 'O';
                if (canWin(true))
                {
                    return;
                }
                field[y][x] = 0;
            }
        }
    }
    int x, y;
    do
    {
        y = rand() % N;
        x = rand() % N;
    }
    while (field[y][x]);
    field[y][x] = 'O';
}

void printField()
{
    std::cout << ' ';
    for (int j = 0 ; j < N ; ++j)
    {
        std::cout << ' ' << (char)('a' + j);
    }
    std::cout << '\n';
    for (int i = 0 ; i < N ; ++i)
    {
        std::cout << i + 1;
        for (int j = 0 ; j < N ; ++j)
        {
            if (field[i][j])
            {
                std::cout << ' ' << field[i][j];
            }
            else
            {
                std::cout << " .";
            }
        }
        std::cout << '\n';
    }
    std::cout << '\n';
}

int main()
{
    srand((int)time(NULL));

    int moves = 0;
    while (true)
    {
        printField();
        if (checkLines())
        {
            std::cout << "Вы выиграли!\n";
            break;
        }
        if (moves == N * N)
        {
            std::cout << "Я победил!\n";
            break;
        }
        playerMove();
        ++moves;
        printField();
        if (checkLines())
        {
            std::cout << "Вы выиграли!\n";
            break;
        }
        std::cout << "Мой ход:\n";
        computerMove();
        ++moves;
    }
}

File: python/tictactoe4.py
#!/usr/bin/python3
import random

N = 4
field = [[None] * N for _ in range(N)]

def playerMove():
    x = -1
    y = -1
    while x < 0 or x >= N or y < 0 or y >= N or field[y][x] is not None:
        s = input('Куда будете ходить? ')
        x = ord(s[0]) - ord('a')
        y = int(s[1]) - 1
    field[y][x] = 'X'

def checkLines():
    for i in range(N):
        found = field[i][0] is not None
        for j in range(N):
            found = found and field[i][j] == field[i][0]
        if found:
            return True
        found = field[0][i] is not None
        for j in range(N):
            found = found and field[j][i] == field[0][i]
        if found:
            return True
    found = field[0][0] is not None
    for j in range(N):
        found = found and field[j][j] == field[0][0]
    return found

def canWin(player):
    if checkLines():
        return False
    for x in range(N):
        for y in range(N):
            if field[y][x] is None:
                field[y][x] = 'X' if player else 'O'
                w = canWin(not player)
                field[y][x] = None
                if not player and w:
                    return True
                if player and not w:
                    return False
    return player

def computerMove():
    for x in range(N):
        for y in range(N):
            if field[y][x] is None:
                field[y][x] = 'O'
                if canWin(True):
                    return
                field[y][x] = None
    while True:
        x = random.randint(0, N - 1)
        y = random.randint(0, N - 1)
        if field[y][x] is None:
            break
    field[y][x] = 'O'

def printField():
    print(' ', end='')
    for j in range(N):
        print(f' {chr(ord("a") + j)}', end='')
    print()
    for i in range(N):
        print(f'{i + 1}', end='')
        for j in range(N):
            if field[i][j] is not None:
                print(f' {field[i][j]}', end='')
            else:
                print(' .', end='')
        print()
    print()

moves = 0
while True:
    printField()
    if checkLines():
        print('Вы выиграли!')
        break
    if moves == N * N:
        print('Я победил!')
        break
    playerMove()
    moves += 1
    printField()
    if checkLines():
        print('Вы выиграли!')
        break
    print('Мой ход:')
    computerMove()
    moves += 1
